common/mam/merkle.c

```diff
@@ -0,0 +1,195 @@
+#include "mam/merkle.h"
+#include "common/sign/v2/iss_curl.h"
+
+static trit_t const merkle_null_hash[HASH_LENGTH] = {0};
+
+// thanks to
+// https://stackoverflow.com/questions/3272424/compute-fast-log-base-2-ceiling#3272516
+static int ceil_log2(unsigned long long x) {
+  static const unsigned long long t[6] = {
+      0xFFFFFFFF00000000ull, 0x00000000FFFF0000ull, 0x000000000000FF00ull,
+      0x00000000000000F0ull, 0x000000000000000Cull, 0x0000000000000002ull};
+
+  int y = (((x & (x - 1)) == 0) ? 0 : 1);
+  int j = 32;
+  int depth_i;
+
+  for (depth_i = 0; depth_i < 6; depth_i++) {
+    int k = (((x & t[depth_i]) == 0) ? 0 : j);
+    y += k;
+    x >>= k;
+    j >>= 1;
+  }
+
+  return y;
+}
+
+static size_t binary_tree_size(size_t depth, size_t acc) {
+  if (depth == 0) {
+    return acc + 1;
+  }
+  return binary_tree_size(depth - 1, acc + (1 << depth));
+}
+
+static size_t merkle_node_base_index(size_t acc, size_t depth, size_t width,
+                                     size_t base_size) {
+  if (depth == 0) return acc;
+  acc += width + 1;
+  if (base_size % 2 == 0)
+    width = width / 2;
+  else
+    width = (width % 2 == 0 ? width / 2 : (width + 1) / 2);
+  return merkle_node_base_index(acc, depth - 1, width, (base_size + 1) / 2);
+}
+
+size_t merkle_node_index(size_t depth, size_t width, size_t base_size,
+                         size_t tree_depth) {
+  size_t acc = 0;
+  size_t idx = width;
+
+  while (depth < tree_depth) {
+    if (width != 0) idx = idx * 2 - 1;
+    acc--;
+    depth++;
+  }
+  return acc + merkle_node_base_index(0, tree_depth, idx, base_size);
+}
+
+size_t merkle_size(size_t leaf_count) {
+  size_t acc = 1;
```

Comment:
here you could just return `leaf_count*2 - 1`
that's because:
pow(2,n) =sum(pow(2,n-1)) + 1
https://math.stackexchange.com/questions/1990137/the-idea-behind-the-sum-of-powers-of-2 
(assuming it is a complete tree)

Reply:
I was indeed assuming it could be an incomplete tree.